python/lsst/daf/butler/core/quantum.py

```diff
@@ -205,70 +158,37 @@ def outputs(self) -> NamedKeyDict[DatasetType, List[DatasetRef]]:
-        """
-        # First validate against predicted
-        if ref.datasetType not in self._predictedInputs:
-            raise ValueError(f"Dataset type {ref.datasetType.name} not in predicted inputs")
-        if ref not in self._predictedInputs[ref.datasetType]:
-            raise ValueError(f"Actual input {ref} was not predicted")
-        # Now insert as actual
-        self._actualInputs.setdefault(ref.datasetType, []).append(ref)
-
-    def addOutput(self, ref: DatasetRef) -> None:
-        """Add an output `DatasetRef` to the `Quantum`.
-
-        This does not automatically update a `Registry`; all `outputs`
-        must be present before a `Registry.addQuantum()` is called.
-
-        Parameters
-        ----------
-        ref : `DatasetRef`
-            Reference for a Dataset to add to the Quantum's outputs.
-        """
-        self._outputs.setdefault(ref.datasetType, []).append(ref)
-
-    @property
-    def id(self) -> Optional[int]:
-        """Unique (autoincrement) integer for this quantum (`int`).
-        """
-        return self._id
-
-    @property
-    def startTime(self) -> Optional[astropy.time.Time]:
-        """Begin timestamp for the execution of this quantum
-        (`astropy.time.Time`).
-        """
-        return self._startTime
-
-    @property
-    def endTime(self) -> Optional[astropy.time.Time]:
-        """End timestamp for the execution of this quantum
-        (`astropy.time.Time`).
-        """
-        return self._endTime
-
-    @property
-    def host(self) -> Optional[str]:
-        """Name of the system on which this quantum was executed (`str`).
-        """
-        return self._host
+        return copy.copy(self._outputs)
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, Quantum):
+            return False
+        for item in ("taskClass", "dataId", "initInputs", "predictedInputs", "outputs"):
+            if getattr(self, item) != getattr(other, item):
+                return False
+        return True
+
+    def __hash__(self) -> int:
+        cumHash = hash(self.taskClass)
+        cumHash ^= hash(self.dataId)
+        if isinstance(self.initInputs, abc.Mapping):
+            values = self.initInputs.values()
+        else:
+            values = self.initInputs
+        for item in values:
+            cumHash ^= hash(item)
+        if self.predictedInputs:
+            for item in self.predictedInputs.values():
+                if isinstance(item, abc.Iterable):
+                    for value in item:
+                        cumHash ^= hash(value)
+                else:
+                    cumHash ^= hash(item)
+        if self.outputs:
+            for item in self.outputs.values():
+                if isinstance(item, abc.Iterable):
+                    for value in item:
+                        cumHash ^= hash(value)
+                else:
+                    cumHash ^= hash(item)
+        return cumHash
```

Comment:
I'm not sure how stable it is to pack all inputs and outputs into a hash (ordering is the obvious question). OTOH it may be enough to use `taskClass` and `dataID` for a hash, in most cases this should be sufficiently unique.

Reply:
That is not enough to compare Quanta prepared from two different input collections. They might have different inputs for instance and those would be considered the same.

Reply:
Well, same hash does not mean object are the same, there is always `__eq__` test involved. I do not think you ever want to put quanta from different graphs into one collection, but even if you do it will still work event if their hashes are equal. My worry is that with a complicated hash it's hard to prove that it is always consistent with `__eq__`.

Reply:
I wrote up a whole thing arguing back and forth based on hash table performance. I got to the end and deleted it all, because now I think you are right.